swim/join_sender.go

```diff
@@ -334,6 +398,68 @@ func (j *joinSender) JoinCluster() ([]string, error) {
 	return nodesJoined, nil
 }
 
+// delayAttempt delays a join attempt by sleeping for an amount of time. The
+// amount of time is computed as an exponential backoff based on the number
+// of join attempts that have been made at the time of the function call;
+// the number of attempts is 0-based. It returns a time.Duration equal to the
+// amount of delay applied and an error if one occurs. If an error occurs, a
+// duration equal to 0 will be returned, signifying no delay has been applied.
+func (j *joinSender) delayAttempt(numAttempts int, randomizer func(int) int,
```

Comment:
we can. I felt as though parameter-injection is a better fit for this function as opposed to injecting them in through the constructor. The dependencies are so specific to this one function that I felt it unnecessary to do hang them off of the `joinSender` type. But, this same thought crossed my mind and I'll try it on for size in the code.